(SERVER-1704) Support flush timeout setting for jruby pool

jpinsonault opened the pull request:
**This PR goes with** a jruby-utils PR which implements the timeout functionality: https://github.com/puppetlabs/jruby-utils/pull/40

---

This commit adds support for a timeout during a jruby pool flush through the
admin api.

Modifies the admin api routes to support returning a 503 if the timeout is
exceeded.

As a bonus it nukes a liberator resource too!

This shouldn't be merged until the corresponding jruby-utils PR is merged and a new release happens

camlow325 commented on src/clj/puppetlabs/services/puppet_admin/puppet_admin_core.clj:
```diff
@@ -49,33 +50,17 @@
+    ; 204 No Content on success
+    {:status 204}
+    ; If a lock timeout occurs return a 503 Service Unavailable
+    (catch [:kind :puppetlabs.services.jruby-pool-manager.impl.jruby-agents/jruby-lock-timeout] e
```
When I run this locally, a timeout doesn't cause anything to be written to the Puppet Server log beyond just the "Draining and refilling" message.  I think it would be good to at least log some sort of error in the case of a flush timeout so it's easy to see from looking at the log when the flush failed.  WDYT?

jpinsonault replied:
Well if it's not a concern then we might as well leave it with the whole namespace to be explicit. Ah yeah good catch, it should log an error

rlinehan commented on test/integration/puppetlabs/services/puppet_admin/puppet_admin_int_test.clj:
```diff
@@ -178,5 +181,26 @@
+                           (str "https://localhost:8140/puppet-admin-api/v1/jruby-pool")
+                           ssl-request-options)]
+             (is (= 503 (:status response)))
+             (is (= "Timeout limit reached before lock could be granted"
```
I think this is going to need to be changed once you update to a released jruby-utils, cuz the string there has changed.... But you'll probably realize that cuz the test will fail :P.

jpinsonault replied:
Thanks for pointing it out either way!

jpinsonault commented on src/clj/puppetlabs/services/puppet_admin/puppet_admin_core.clj:
```diff
@@ -6,10 +6,12 @@
             [puppetlabs.puppetserver.liberator-utils :as liberator-utils]
             [schema.core :as schema]
             [liberator.core :refer [defresource]]
-    ;[liberator.dev :as liberator-dev]
```
Ah, ok, well I'm fine with leaving it

jpinsonault commented on test/integration/puppetlabs/services/puppet_admin/puppet_admin_int_test.clj:
```diff
@@ -178,5 +181,26 @@
+             borrowed-instance (.borrowItem pool)]
+         (try
+           (let [response (http-client/delete
+                           (str "https://localhost:8140/puppet-admin-api/v1/jruby-pool")
```
derp, copy paste oversight

jpinsonault commented:
Once https://github.com/puppetlabs/jruby-utils/pull/41 is merged I'll do a release and push another set of commits

jpinsonault commented:
Addressed the last set of feedback, updated jruby-utils to the released version, and rebased to change the commit message to explain why the liberator resource was removed

rlinehan commented:
🎈 🎉
